go/vt/mysqlctl/backup.go

```diff
@@ -0,0 +1,563 @@
+	// - backupInnodbLogGroupHomeDir for files that go into Mycnf.InnodbLogGroupHomeDir
+	// - backupData for files that go into Mycnf.DataDir
+	Base string
+
+	// Name is the file name, relative to Base
+	Name string
+
+	// Hash is the hash of the gzip compressed data stored in the
+	// BackupStorage.
+	Hash string
+}
+
+func (fe *FileEntry) open(cnf *Mycnf, readOnly bool) (*os.File, error) {
+	// find the root to use
+	var root string
+	switch fe.Base {
+	case backupInnodbDataHomeDir:
+		root = cnf.InnodbDataHomeDir
+	case backupInnodbLogGroupHomeDir:
+		root = cnf.InnodbLogGroupHomeDir
+	case backupData:
+		root = cnf.DataDir
+	default:
+		return nil, fmt.Errorf("unknown base: %v", fe.Base)
+	}
+
+	// and open the file
+	name := path.Join(root, fe.Name)
+	var fd *os.File
+	var err error
+	if readOnly {
+		fd, err = os.Open(name)
+	} else {
+		dir := path.Dir(name)
+		if err := os.MkdirAll(dir, os.ModePerm); err != nil {
+			return nil, fmt.Errorf("cannot create destination directory %v: %v", dir, err)
+		}
+		fd, err = os.Create(name)
+	}
+	if err != nil {
+		return nil, fmt.Errorf("cannot open source file %v: %v", name, err)
+	}
+	return fd, nil
+}
+
+// BackupManifest represents the backup. It lists all the files, and
+// the ReplicationPosition that the backup was taken at.
+type BackupManifest struct {
+	// FileEntries contains all the files in the backup
+	FileEntries []FileEntry
+
+	// ReplicationPosition is the position at which the backup was taken
+	ReplicationPosition proto.ReplicationPosition
+}
+
+// isDbDir returns true if the given directory contains a DB
+func isDbDir(p string) bool {
+	// db.opt is there
+	if _, err := os.Stat(path.Join(p, "db.opt")); err == nil {
+		return true
+	}
+
+	// Look for at least one .frm file
+	fis, err := ioutil.ReadDir(p)
+	if err != nil {
+		return false
+	}
+	for _, fi := range fis {
+		if strings.HasSuffix(fi.Name(), ".frm") {
+			return true
+		}
+	}
+
+	return false
+}
+
+func addDirectory(fes []FileEntry, base string, baseDir string, subDir string) ([]FileEntry, error) {
+	p := path.Join(baseDir, subDir)
+
+	fis, err := ioutil.ReadDir(p)
+	if err != nil {
+		return nil, err
+	}
+	for _, fi := range fis {
+		fes = append(fes, FileEntry{
+			Base: base,
+			Name: path.Join(subDir, fi.Name()),
+		})
+	}
+	return fes, nil
+}
+
+func findFilesTobackup(cnf *Mycnf, logger logutil.Logger) ([]FileEntry, error) {
+	var err error
+	var result []FileEntry
+
+	// first add inno db files
+	result, err = addDirectory(result, backupInnodbDataHomeDir, cnf.InnodbDataHomeDir, "")
+	if err != nil {
+		return nil, err
+	}
+	result, err = addDirectory(result, backupInnodbLogGroupHomeDir, cnf.InnodbLogGroupHomeDir, "")
+	if err != nil {
+		return nil, err
+	}
+
+	// then add DB directories
+	fis, err := ioutil.ReadDir(cnf.DataDir)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, fi := range fis {
+		p := path.Join(cnf.DataDir, fi.Name())
+
+		// If this is not a directory, try to eval it as a syslink.
```

Comment:
Honestly I don't know, I copied the old code without trying it out. I'll add a unit test case. The idea is to allow symlinks for the db directories so they can be on different mounted physical partitions if required by the setup.